### ISSUE TITLE
What's a good practice for data versioning and backward compatibility?

### ISSUE THREAD

**Issue by YoniH** [Netflix/falcor#831]
In Falcor, the data is the API, and according to the docs, a common pattern to build apps is that the view retrieves data directly from the model.
That means that the view knows the structure of the JSON Graph. But what if you want to make changes in the structure?
For web clients it might not be a big problem, because refreshing the page would get the user the newest client code. But mobile clients are already out and might not be updated, so you still want to support the old structure.
In a REST api, you would create a new API version, keeping the old versions for backward compatibility.
How would you do it in Falcor? Keep versions of the graph?

**Comment by eddieajau** [comment]
My advice is don't, ever make backward incompatible changes to your routes. Only ever add routes and leave the old routes in play while you are waiting for deprecated dependencies to die.

For example, say your first stab at an Article model is `GET articlesById[1].name`.

That's now in production. Now let's say that you want to change `name` to title, ie `GET articlesById[1].title`.

I would keep _both_ published in Falcor. So using your question about mobile applications.

v1 of your mobile build will look for the Article `name`.

Now you upgrade your Falcor server and also cut v2 of your mobile app which expects an Article `title`.

A person upgrades to v2 and has no trouble accessing Article _titles_. The v2 app will also be away of any new paths you have added.

A person who has not yet upgraded can still use the v1 app and see Article _names_.

In this way you can insulate breaking changes in your underlying data sources using Falcor (for example, both `name` and `title` are coming from the same backend data source).

Does that make sense? Does that help?

In other words, don't version your JSON Graph, only deprecate.

**Comment by mattiaerre** [comment]
ditto @eddieajau 😊

**Comment by Joshfindit** [comment]
That approach has failed for others during this common scenario:

1. API is created with a small bug
2. Users start relying on the bug (possibly without reporting it)
3. Later the bug is found and 'breaking' API changes need to be made

Thoughts?

**Comment by mattiaerre** [comment]
Hi @Joshfindit ,
I would try to add the "fixed" data against a new structure and maybe return an error object instead of a normal value as a response when the user wants to consume the "broken" structure. Inside this error type, you can tell which new structure to consume. Just an idea.

**Comment by eddieajau** [comment]
@Joshfindit can you give a bit of an example?

**Comment by trxcllnt** [comment]
You should be able to version your `model.json` endpoint by versioning the URL. The route `yourapp.com/api/v1.0.0/model.json` can serve the v1.0.0 routes, and any new backwards-incompatible routes can be stood up at `yourapp.com/api/v2.0.0/model.json`. This lets you stagger the the roll-out of different clients with different API versions, without breaking any along the way.

**Comment by mattiaerre** [comment]
this is really good @trxcllnt , would you recommend to use your approach? maybe cleaner.
thanks.

**Comment by greim** [comment]
@trxcllnt gives a good suggestion. I like of the idea of never deprecating an old route, but it seems overly-restrictive as a blanket rule. Another, possibly wacky, idea is having the version number as a step in the path.

```
v1.articles[0..19].name
v2.articles[0..19].title
```

This would give you the option of aliasing across the two, so that the a front-end could migrate from one API to another without duplicating data across two models.

**Comment by eddieajau** [comment]
> I like of the idea of never deprecating an old route, but it seems overly-restrictive as a blanket rule.

Not really. You still have to answer the question "when can I turn off this old API?". It's the same problem whether you use v's in the URL or path, or whether you maintain a fully backward compatible API.

That said, it may be worth putting some thought into a permanent API to be able to tell a device whether the App it's using is still supported. Thinking out loud, `GET app.meta.minVersion` might be able to prevent you from using the App if it's too old.

**Comment by steveorsomethin** [comment]
I'm currently performing triage on the falcor issues, and closing/tagging as I go.

Paul provided the same answer I would. Closing.
